src/at_tools/att_doc_sources.rs

```diff
@@ -0,0 +1,374 @@
+                .map_err(|e| format!("Unable to create directory {:?} {directory:?}: {e:?}", env::current_dir()))?;
+            let mut file = File::create(&file_path)
+                .map_err(|_| format!("Unable to create file {file_path}"))?;
+            file.write_all(text.as_bytes())
+                .map_err(|_| format!("Unable to write to file {file_path}"))?;
+            pages.insert(url, file_path.clone());
+
+            // vectorize file
+            let gcx = gcx.write().await;
+            let vec_db_module = {
+                *gcx.documents_state.cache_dirty.lock().await = true;
+                gcx.vec_db.clone()
+            };
+            let document = Document::new(&PathBuf::from(file_path.as_str()));
+            match *vec_db_module.lock().await {
+                Some(ref mut db) => db.vectorizer_enqueue_files(&vec![document], false).await,
+                None => {}
+            };
+
+            if is_last_iteration {
+                continue;
+            }
+
+            // find links
+            let base_parser = Url::options().base_url(Some(&base_url));
+            select::document::Document::from(html.as_str())
+                .find(Name("a"))
+                .filter_map(|n| n.attr("href"))
+                .for_each(|link| {
+                    let Ok(link) = base_parser.parse(link) else {
+                        return;
+                    };
+                    let link = link.as_str();
+                    if !link.starts_with(&base_url.to_string()) {
+                        return;
+                    }
+                    if visited_pages.len() >= max_pages || visited_pages.contains(link) {
+                        return;
+                    }
+                    new_urls.push(link.to_string());
+                    visited_pages.insert(link.to_string());
+                });
+        }
+
+        queue = vec![];
+        swap(&mut queue, &mut new_urls);
+    }
+
+    let (directory, _) = get_directory_and_file_from_url(&url_str).unwrap();
+
+    let origin = DocOrigin {
+        url: url_str.clone(),
+        max_depth: depth,
+        max_pages,
+        pages,
+    };
+
+    if let Ok(origin_json) = serde_json::to_string(&origin) {
+        let file_path = format!("./.refact/docs/{directory}/origin.json");
+        let mut file = File::create(&file_path)
+            .map_err(|_| format!("Unable to create file {file_path}"))?;
+        file.write_all(origin_json.as_bytes())
+            .map_err(|_| format!("Unable to write to file {file_path}"))?;
+    } else {
+        warn!("Unable to convert DocOrigin to json");
+    }
+
+    Ok(sources)
+}
+
+async fn doc_sources_list(ccx: &mut AtCommandsContext, tool_call_id: &String) -> Result<Vec<ContextEnum>, String> {
+    let sources = ccx
+        .global_context
+        .read()
+        .await
+        .documents_state
+        .documentation_sources
+        .lock()
+        .await
+        .join(",");
```

Comment:
Personally, I'm not a fan of this writing style and we don't write in a project like this, why does it take 8 lines when it could be one single line?